src/zeroconf/_services/info.py

```diff
@@ -355,57 +367,67 @@ def async_update_records(self, zc: 'Zeroconf', now: float, records: List[RecordU
 
         This method will be run in the event loop.
         """
-        self._process_records_threadsafe(zc, now, records)
+        if self._process_records_threadsafe(zc, now, records) and self._notify_event:
+            self._notify_event.set()
+            self._notify_event.clear()
 
-    def _process_records_threadsafe(self, zc: 'Zeroconf', now: float, records: List[RecordUpdate]) -> None:
-        """Thread safe record updating."""
+    def _process_records_threadsafe(self, zc: 'Zeroconf', now: float, records: List[RecordUpdate]) -> bool:
+        """Thread safe record updating.
+
+        Returns True if the service info was updated.
+        """
         seen_addresses: Set[bytes] = set()
+        updated: bool = False
         for record_update in records:
             record = record_update.new
             if isinstance(record, DNSAddress):
                 seen_addresses.add(record.address)
-            self._process_record_threadsafe(record, now)
+            updated |= self._process_record_threadsafe(record, now)
         for record in self._get_address_records_from_cache(zc):
             if record.address not in seen_addresses:
-                self._process_record_threadsafe(record, now)
+                updated |= self._process_record_threadsafe(record, now)
+        return updated
 
-    def _process_record_threadsafe(self, record: DNSRecord, now: float) -> None:
+    def _process_record_threadsafe(self, record: DNSRecord, now: float) -> bool:
         """Thread safe record updating."""
         if record.is_expired(now):
-            return
+            return False
 
-        if isinstance(record, DNSAddress):
-            if record.key != self.server_key:
-                return
+        if record.key == self.server_key and isinstance(record, DNSAddress):
             try:
                 ip_addr = _cached_ip_addresses(record.address)
             except ValueError as ex:
                 log.warning("Encountered invalid address while processing %s: %s", record, ex)
-                return
-            if isinstance(ip_addr, ipaddress.IPv4Address):
+                return False
+            if ip_addr.version == 4:
                 if ip_addr not in self._ipv4_addresses:
                     self._ipv4_addresses.insert(0, ip_addr)
-                return
+                    return True
+                return False
             if ip_addr not in self._ipv6_addresses:
                 self._ipv6_addresses.insert(0, ip_addr)
                 if ip_addr.is_link_local:
                     self.interface_index = record.scope_id
```

Comment:
There is a bug here. If its not the first address in the list we need to pop it back out